### PR TITLE
Finalising Safari Port

### PR DESCRIPTION
**Pull request by heychirag**
I have systematically compared the entire folder structure of chrome and safari using [meld](https://en.wikipedia.org/wiki/Meld_(software)). There were many files between chrome and safari that only differed by a few functions or maybe only a few lines of code. Therefore, I decided to add a snippet in files that would detect the browser and then run the specific API. This however might not look like a good coding practice but in the long run it would simplify the maintenance and understandability of the codebase. I could have followed Kolya's method but Safari's documentation simply doesn't allows us to do so.

### CODE REVIEW

**Comment by Ceilican** [comment]
Hi @heychirag ! Travis's CI tests are failing. Could you fix this before I merge the PR?

File: Source/Chrome/background/background.js [hunk 1]
```diff
@@ -7,75 +7,257 @@
...
-
 /**
  * namespace for background
  */
 
 var background = {};
 
 /**
- * script element for xml2json
+ * check if the browser is chrome or safari
```
Comment (Ceilican) [inline Source/Chrome/background/background.js]: What about Firefox?

File: Source/Core/pages/js/browserServices.js [hunk 2]
```diff
@@ -7,104 +7,116 @@ browserServices.checkBrowser = function() {
...
 	if ('firefoxApi' in window)
 		browser = 'firefox';
-	if ('safariApi' in window)
-		browser = 'safari';
-
+  /**
+   * detecting SAFARI browser
+   * 'chrom' filters down (chrom)e as well as (chrom)ium 
+   */
+  if(!(navigator.userAgent.toLowerCase().indexOf("chrom") != -1)){
```
Comment (Ceilican) [inline Source/Core/pages/js/browserServices.js]: what about Firefox?

Comment (Kolpa) [inline Source/Core/pages/js/browserServices.js]: Firefox uses a different background page than this so it is decided earlyer down the Road

File: Source/Chrome/background/background.js (hunk 1 above)
Comment (Kolpa) [inline Source/Chrome/background/background.js]: Firefox uses a different background page than this so it is decided earlyer down the Road

**Comment by Ceilican** [comment]
> Therefore, I decided to add a snippet in files that would detect the browser and then run the specific API. This however might not look like a good coding practice but in the long run it would simplify the maintenance and understandability of the codebase.

I had a superficial look on the code. I could see that, in some cases, we have a uniform treatment of the three browsers. But, in some cases, we just check between Safari and Chrome, apparently forgetting to consider Firefox. Therefore, I am not 100% confident that this code will work. Maybe it will work in Safari, but will break the Firefox port. I haven't had time to check this, though.  (@Kolpa , since you were responsible for the Firefox port, could you check whether the Firefox port would still work with this PR? )

>  I could have followed Kolya's method but Safari's documentation simply doesn't allows us to do so.

Could you explain this in more detail? Why does the Safari API doesn't allow us to treat Safari in the same way as Chrome and Firefox were treated?

Cheers!

**Comment by Kolpa** [comment]
I'll have a look into Firefox compatibility of this pr tomorrow it is 3 am right now and I should really catch some sleep but I can Shine a light on how the Firefox Port works differently for now.

The basic difference is the Architecture of the Plugins,  Firefox is more Focused on a background System Kind of Plugin and Chrome and Safari basically share the content Script intensive approach. To fix These issuis and allow Firefox to work i wrote Wrapper apis for all Interaction with the old Chrome apis and then check dependant on the current Browser if we should invoke the Chrome/Safari apis or contact the Firefox Backend and wait for the data from there. This is done by loading a custom content Script inside the options page when running on Firefox which exposes a few functions to contact the Firefox Backend and if these are not there it is assumed that we are running on Chrome or Safari. I hope this sheds some light on the Situation, feel free to ask if you need to know anything else.

**Comment by Ceilican** [comment]
Thanks for the explanation, @Kolpa . 

I am positively surprised to learn that Safari's API is architecturally more similar to Chrome's API than Firefox's API. 

@heychirag: considering these architectural differences, it seem that you chose a good approach.

I propose the following plan:

1) tomorrow (Tuesday, India time), @heychirag fixes the Travis-CI test failure (in case it is caused by this PR. Otherwise, if it was caused by someone else, let me know.)

2) tomorrow (Tuesday, Germany time), @Kolpa checks that the Firefox port still works with this PR.

3) on wednesday (Australia time, probably in the evening), I will check both the Firefox port and the Safari port at the same time, and I will have a closer look on the porting solutions in our code in general.

**Comment by heychirag** [comment]
Hi @Ceilican,

The commit is passing the Travis CI tests now. Now for your concern regarding negligence to check for Firefox I can assure you that it won't break Firefox's functionality. Following are the navigator.userAgent strings as returned by different browsers.

Chrome returns:
`User-agent header sent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36`

Safari returns:
`User-agent header sent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12) AppleWebKit/602.1.50 (KHTML, like Gecko) Version/10.0 Safari/602.1.50`

Firefox returns:
`User-agent header sent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:45.0) Gecko/20100101 Firefox/45.0`

Therefore to check if the browser is Safari, we first need to make sure that it isn't Chrome as Chrome also returns Safari in it's userAgent string. Though all browsers return Mozilla but Firefox returns neither Chrome nor Safari. But again, we are not checking for Firefox since Kolya has implemented a different approach for it.

**Comment by Kolpa** [comment]
as of right now there are no issiues with this commit in firefox.

**Comment by Ceilican** [comment]
Thanks, @heychirag ! I can see now that the tests are passing. However, there are conflicts preventing automatic merging of this PR. Could you check that?

**Comment by heychirag** [comment]
@Ceilican, you can merge the PR now.

**Comment by Ceilican** [comment]
Thanks!!

**Pull request merged**